### PR TITLE
Update yoda gas estimation function

### PR DESCRIPTION
**Pull request by thebevrishot**
## Implementation details

- Update yoda gas estimation logic

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [x] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: chain/yoda/execute.go [hunk 1]
```diff
@@ -121,7 +87,7 @@ func SubmitReport(c *Context, l *Logger, keyIndex int64, reports []ReportMsgWith
 	memo := fmt.Sprintf("yoda:%s/exec:%s", version.Version, strings.Join(versions, ","))
 	key := c.keys[keyIndex]
 	cliCtx := sdkCtx.CLIContext{Client: c.client, TrustNode: true, Codec: cdc}
-	gasLimit := estimatedReportsGas(msgs)
+	gasLimit := estimateGas(*c, msgs, feeEstimations)
```
Comment (taobun) [inline chain/yoda/execute.go]: Send context as pointer like other function.

File: chain/yoda/gas.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	countingReportCost = uint64(1479)
+	addingPendingCost  = uint64(3400)
+
+	baseRequestSize = uint64(32)
+	addressSize     = uint64(20)
+
+	baseRawRequestSize = uint64(16)
+)
+
+func estimateTxSize(c Context, msgs []sdk.Msg) uint64 {
```
Comment (taobun) [inline chain/yoda/gas.go]: Is context used?

File: chain/yoda/gas.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		}
+
+		ser := cdc.MustMarshalBinaryBare(msg)
+		size += uint64(len(ser))
+	}
+
+	return size
+}
+
+func estimateStoringReportCost(c Context, msg sdk.Msg) uint64 {
```
Comment (taobun) [inline chain/yoda/gas.go]: ditto

File: chain/yoda/gas.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	cost := writingBaseCost
+	cost += uint64(len(cdc.MustMarshalBinaryBare(msg))) * writingCostPerByte
```
Comment (taobun) [inline chain/yoda/gas.go:53]: This is the estimated cost of report in storage not a real one right?

File: chain/yoda/context.go [hunk 5]
```diff
@@ -13,10 +13,19 @@ import (
 	"github.com/bandprotocol/bandchain/chain/yoda/executor"
 )
 
+type FeeEstimationData struct {
+	minCount    int64
+	callData    []byte
+	validators  []string
```
Comment (taobun) [inline chain/yoda/context.go]: We used only size of the validator?

File: chain/yoda/gas.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	cost += addingPendingCost
+
+	return cost
+}
+
+func estimateGas(c Context, msgs []sdk.Msg, feeEstimations []FeeEstimationData) uint64 {
+	gas := baseFixedGas
+
+	txSize := estimateTxSize(c, msgs)
+	gas += txCostPerByte * txSize
```
Comment (taobun) [inline chain/yoda/gas.go]: Combine to 1 line looks better I think.

File: chain/yoda/gas.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+	readingBaseCost = uint64(1000)
+	writingBaseCost = uint64(2000)
+
+	readingCostPerByte = uint64(3)
+	writingCostPerByte = uint64(30)
+
+	// band
+	baseReportCost     = uint64(4006)
+	payingFeeCost      = uint64(15200)
```
Comment (taobun) [inline chain/yoda/gas.go]: Move it to cosmos section

File: chain/yoda/gas.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	countingReportCost = uint64(1479)
```
Comment (taobun) [inline chain/yoda/gas.go]: countingPerReportCost

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/yoda/gas.go (hunk 4 above)
Comment (thebevrishot) [reply chain/yoda/gas.go:53]: This is real value that would be stored.

File: chain/yoda/context.go (hunk 5 above)
Comment (thebevrishot) [reply chain/yoda/context.go]: yes, that's make sense. At first I think size of validator addresses might be variant.

File: chain/yoda/gas.go (hunk 6 above)
Comment (thebevrishot) [reply chain/yoda/gas.go]: It would be easier while reading code.

File: chain/yoda/gas.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+	writingBaseCost = uint64(2000)
+
+	readingCostPerByte = uint64(3)
+	writingCostPerByte = uint64(30)
+
+	payingFeeCost = uint64(15200)
+
+	// band
+	baseReportCost        = uint64(4006)
+	countingPerReportCost = uint64(1479)
```
Comment (taobun) [inline chain/yoda/gas.go]: Using report size to calculate this.

File: chain/yoda/context.go [hunk 10]
```diff
@@ -13,10 +13,19 @@ import (
 	"github.com/bandprotocol/bandchain/chain/yoda/executor"
 )
 
+type FeeEstimationData struct {
+	minCount    int64
```
Comment (taobun) [inline chain/yoda/context.go:18]: using askCount instead.

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/yoda/handler.go [hunk 11]
```diff
@@ -158,22 +164,23 @@ func handleRequestLog(c *Context, l *Logger, log sdk.ABCIMessageLog) {
...
 		execVersion: execVersions,
 		keyIndex:    keyIndex,
 		feeEstimationData: FeeEstimationData{
+			askCount:    askCount,
 			minCount:    minCount,
 			callData:    callData,
 			validators:  len(validators),
 			rawRequests: reqs,
-			clientId:    clientId,
+			clientId:    clientID,
```
Comment (taobun) [inline chain/yoda/handler.go]: clientID!!!

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**